source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,624 @@
+        xConnectInfo.pcPassword = xConnectInfo.pcUserName + azureiotUSERNAME_MAX;
+
+        if( az_result_failed( xCoreResult = az_iot_hub_client_get_user_name( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                             ( char * ) xConnectInfo.pcUserName, azureiotUSERNAME_MAX,
+                                                                             &xMQTTUserNameLength ) ) )
+        {
+            AZLogError( ( "Failed to get username: 0x%08x.", xCoreResult ) );
+            xResult = eAzureIoTHubClientFailed;
+        }
+        /* Check if token refresh is set, then generate password */
+        else if( ( pxAzureIoTHubClient->_internal.pxAzureIoTHubClientTokenRefresh ) &&
+                 ( pxAzureIoTHubClient->_internal.pxAzureIoTHubClientTokenRefresh( pxAzureIoTHubClient,
+                                                                                   pxAzureIoTHubClient->_internal.xAzureIoTHubClientTimeFunction() +
+                                                                                   azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC,
+                                                                                   pxAzureIoTHubClient->_internal.pucAzureIoTHubClientSymmetricKey,
+                                                                                   pxAzureIoTHubClient->_internal.ulAzureIoTHubClientSymmetricKeyLength,
+                                                                                   ( uint8_t * ) xConnectInfo.pcPassword, azureiotPASSWORD_MAX,
+                                                                                   &ulPasswordLength ) ) )
+        {
+            AZLogError( ( "Failed to generate SAS token." ) );
+            xResult = eAzureIoTHubClientFailed;
+        }
+        else
+        {
+            xConnectInfo.ucCleanSession = xCleanSession;
+            xConnectInfo.pcClientIdentifier = ( const char * ) pxAzureIoTHubClient->_internal.pucDeviceID;
+            xConnectInfo.usClientIdentifierLength = ( uint16_t ) pxAzureIoTHubClient->_internal.ulDeviceIDLength;
+            xConnectInfo.usUserNameLength = ( uint16_t ) xMQTTUserNameLength;
+            xConnectInfo.usKeepAliveSeconds = azureiothubKEEP_ALIVE_TIMEOUT_SECONDS;
+            xConnectInfo.usPasswordLength = ( uint16_t ) ulPasswordLength;
+
+            /* Send MQTT CONNECT packet to broker. Last Will and Testament is not used. */
+            if( ( xMQTTResult = AzureIoTMQTT_Connect( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                      &xConnectInfo,
+                                                      NULL,
+                                                      ulTimeoutMilliseconds,
+                                                      &pxOutSessionPresent ) ) != eAzureIoTMQTTSuccess )
```

Comment:
>pxOutSessionPresent [](start = 55, length = 19)

not sure why you need double pointer

Reply:
and type is uint8_t

---
In reply to: [610890192](https://github.com/Azure/azure-iot-middleware-freertos/pull/60#discussion_r610890192) [](ancestors = 610890192)

Reply:
oops yea find all and replace error here